lib/commands/apps-management.js

```diff
@@ -0,0 +1,94 @@
+import _ from 'lodash';
+import waitForCondition from 'asyncbox';
+import {fs, utils} from 'appium-support';
+import log from '../logger';
+
+let commands = {};
+const APP_STATE_NOT_INSTALLED = 0;
+const APP_STATE_NOT_RUNNING = 1;
+const APP_STATE_RUNNING_IN_FOREGROUND = 2;
+const APP_STATE_RUNNING_IN_BACKGROUND = 3;
+
+
+function extractMandatoryOptions (opts = {}, namesList) {
+  const result = {};
+  for (const name of namesList) {
+    if (!utils.hasValue(opts[name])) {
+      throw new Error(`'${name}' option is mandatory.`);
+    }
+    result[name] = opts[name];
+  }
+  return result;
+}
+
+commands.mobileInstallApp = async function (opts = {}) {
```

Comment:
don't we already have logic for installing / uninstalling apps, either in the driver or in appium-adb? i wouldn't want to duplicate it unless necessary

Reply:
we have some parts of this logic here and there, but not everything in one place.

Reply:
ok makes sense. my other question is, why make these "mobile" methods when appium endpoints already exist?

Reply:
These endpoints work with the application under test only and won't allow one to handle any other package on the device including built-in packages

Reply:
why does this method take an 'app' option then, if it is only to be used to install the AUT, whose info we presumably already have?

Reply:
it can be a path to any apk or an URL. Don't you like the argument name?

Reply:
I think I'm just confused about the point of this. It looked to me like you were adding a new command for clients, like:

```js
await driver.execute("mobile: installApp", {app: '/path/to/my.apk'});
```

in the context of an existing session, why would someone use this to install anything other than a separate app? if the point is to do a reinstall only, should we enforce same package as AUT? i'm just trying to make sure this is different from the existing endpoint.

Reply:
One of possible scenarios - we need to deploy some helper app for the automation along with the main AUT. Reinstall and upgrade are also possible usage scenarios.

Reply:
why would someone not just use `driver.installApp` for those purposes as well?

Reply:
because it affects the whole session